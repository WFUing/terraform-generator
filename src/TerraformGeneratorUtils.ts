import replaceString from 'replace-string';
import { TerraformVersion, Block, Argument, Map } from '.';

export type ArgumentsToStringFn = (version: TerraformVersion, key: string, value: any) => string;

export default class TerraformGeneratorUtils {

  private static readonly escapeChars = [
    ['"', '&tfgquot;']
  ];

  static escape(str: string): string {
    this.escapeChars.forEach(char => str = replaceString(str, char[0], char[1]));
    return str;
  }

  static unescape(str: string): string {
    this.escapeChars.slice().reverse().forEach(char => str = replaceString(str, char[1], char[0]));
    return str;
  }

  static argumentsToString(version: TerraformVersion, args: Record<string, any>, customArgumentToString?: ArgumentsToStringFn): string {
    let str = '';
    for (const key in args) {
      if (customArgumentToString) {
        str += customArgumentToString(version, key, args[key]);
      } else {
        str += this.argumentToString(version, key, args[key]);
      }
    }
    return str;
  }

  static isObjectArgument(value: any): boolean {
    if (['string', 'number', 'boolean'].indexOf(typeof value) >= 0
      || value instanceof Block || value instanceof Argument || value instanceof Map) {
      return false;

    } else if (typeof value === 'object') {
      return true;

    } else {
      throw new Error(`Invalid value: ${value}`);
    }
  }

  static argumentToString(version: TerraformVersion, key: string, value: any): string {
    try {
      if (value == null) {
        return '';
      }

      if (key.indexOf(' ') > 0) {
        key = `"${key}"`;
      }

      let operator = ' = ';
      let isObjectArray = false;

      if (Array.isArray(value)) {
        if (value.length === 0 || this.isObjectArgument(value[0])) {
          value = value.filter(element => element != null);
          operator = ' ';
          isObjectArray = true;
        }
      } else {
        if (this.isObjectArgument(value)) {
          operator = ' ';
        }
      }

      if (isObjectArray) {
        let str = '';
        if (Array.isArray(value)) {
          value.forEach(element => {
            str += `${key}${operator}${this.argumentValueToString(version, element)}\n`;
          });
        }
        return str;

      } else {
        return `${key}${operator}${this.argumentValueToString(version, value)}\n`;
      }

    } catch (err) {
      throw new Error(`Invalid value: ${key} = ${value}`);
    }
  }

  static argumentValueToString(version: TerraformVersion, value: any): string {
    if (value == null) {
      return null;

    } else if (value instanceof Block) {
      return this.argumentValueToString(version, value.asArgument());

    } else if (value instanceof Argument) {
      if (version && version === '0.11' && !value.asIs) {
        return `"\${${value.toTerraform()}}"`;
      } else {
        return value.toTerraform();
      }

    } else if (value instanceof Map) {
      return this.argumentValueToString(version, value.arguments);

    } else if (['string', 'number', 'boolean'].indexOf(typeof value) >= 0) {
      return JSON.stringify(value);

    } else if (typeof value === 'object') {
      if (Array.isArray(value)) {
        let str = '[\n';
        value.forEach((element, i) => {
          str += `${this.argumentValueToString(version, element)}${i < value.length - 1 ? ',' : ''}\n`;
        });
        str += ']';
        return str;

      } else {
        let str = '{\n';
        for (const key in value) {
          str += this.argumentToString(version, key, value[key]);
        }
        str += '}';
        return str;
      }

    } else {
      throw new Error(`Invalid value: ${value}`);
    }
  }

}
