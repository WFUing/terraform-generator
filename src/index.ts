export { BlockArgs } from './utils';

export { TerraformGenerator, WriteOptions } from './TerraformGenerator';

export {
  Block,
  Backend,
  Comment,
  Data,
  Import, ImportArgs,
  Locals,
  Module, ModuleArgs,
  Output, OutputArgs,
  Provider,
  Provisioner, ProvisionerArgs, ProvisionerType,
  Resource, ResourceToDataOptions,
  Variable, VariableArgs
} from './blocks';

export {
  Argument, arg,
  Attribute, attr,
  Function, fn,
  Heredoc, heredoc
} from './arguments';

export {
  List, list,
  Map, map
} from './types';
