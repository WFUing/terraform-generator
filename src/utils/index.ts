export * from './types';
export * from './Util';
