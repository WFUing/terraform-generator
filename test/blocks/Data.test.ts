import { Data } from '../../src';
import { arg4 } from '..';

test('Data', () => {
  const data = new Data('type', 'name', arg4);
  expect(data.toTerraform('0.11')).toMatchSnapshot();
  expect(data.toTerraform('0.12')).toMatchSnapshot();
  expect(data.asArgument().toTerraform()).toBe('data.type.name');
  expect(data.attr('attr').toTerraform()).toBe('data.type.name.attr');
});
