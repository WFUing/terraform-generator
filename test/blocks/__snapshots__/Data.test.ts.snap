// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data 1`] = `
"data \\"type\\" \\"name\\"{
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = \\"\${type.name.attr}\\"
arg7 = [
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
]
arg8 = \\"\${type.name}\\"
arg9 = [
\\"\${type.name}\\",
\\"\${type.name}\\",
\\"\${type.name}\\",
]
arg10 = [
\\"e\\",
4,
true,
\\"\${type.name.attr}\\",
\\"\${type.name}\\",
]
arg11 = \\"\${arg}\\"
arg12 = \\"\${type.name.attr}\\"
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = \\"\${fn(1, 2, 3)}\\"
arg16 = \\"\${fn(type.name.attr, type.name.attr, type.name.attr)}\\"
arg100 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = \\"\${type.name.attr}\\"
arg7 = [
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
]
arg8 = \\"\${type.name}\\"
arg9 = [
\\"\${type.name}\\",
\\"\${type.name}\\",
\\"\${type.name}\\",
]
arg10 = [
\\"e\\",
4,
true,
\\"\${type.name.attr}\\",
\\"\${type.name}\\",
]
arg11 = \\"\${arg}\\"
arg12 = \\"\${type.name.attr}\\"
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = \\"\${fn(1, 2, 3)}\\"
arg16 = \\"\${fn(type.name.attr, type.name.attr, type.name.attr)}\\"
}
arg101 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = \\"\${type.name.attr}\\"
arg7 = [
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
]
arg8 = \\"\${type.name}\\"
arg9 = [
\\"\${type.name}\\",
\\"\${type.name}\\",
\\"\${type.name}\\",
]
arg10 = [
\\"e\\",
4,
true,
\\"\${type.name.attr}\\",
\\"\${type.name}\\",
]
arg11 = \\"\${arg}\\"
arg12 = \\"\${type.name.attr}\\"
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = \\"\${fn(1, 2, 3)}\\"
arg16 = \\"\${fn(type.name.attr, type.name.attr, type.name.attr)}\\"
arg100 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = \\"\${type.name.attr}\\"
arg7 = [
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
]
arg8 = \\"\${type.name}\\"
arg9 = [
\\"\${type.name}\\",
\\"\${type.name}\\",
\\"\${type.name}\\",
]
arg10 = [
\\"e\\",
4,
true,
\\"\${type.name.attr}\\",
\\"\${type.name}\\",
]
arg11 = \\"\${arg}\\"
arg12 = \\"\${type.name.attr}\\"
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = \\"\${fn(1, 2, 3)}\\"
arg16 = \\"\${fn(type.name.attr, type.name.attr, type.name.attr)}\\"
}
}
arg101 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = \\"\${type.name.attr}\\"
arg7 = [
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
]
arg8 = \\"\${type.name}\\"
arg9 = [
\\"\${type.name}\\",
\\"\${type.name}\\",
\\"\${type.name}\\",
]
arg10 = [
\\"e\\",
4,
true,
\\"\${type.name.attr}\\",
\\"\${type.name}\\",
]
arg11 = \\"\${arg}\\"
arg12 = \\"\${type.name.attr}\\"
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = \\"\${fn(1, 2, 3)}\\"
arg16 = \\"\${fn(type.name.attr, type.name.attr, type.name.attr)}\\"
arg100 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = \\"\${type.name.attr}\\"
arg7 = [
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
]
arg8 = \\"\${type.name}\\"
arg9 = [
\\"\${type.name}\\",
\\"\${type.name}\\",
\\"\${type.name}\\",
]
arg10 = [
\\"e\\",
4,
true,
\\"\${type.name.attr}\\",
\\"\${type.name}\\",
]
arg11 = \\"\${arg}\\"
arg12 = \\"\${type.name.attr}\\"
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = \\"\${fn(1, 2, 3)}\\"
arg16 = \\"\${fn(type.name.attr, type.name.attr, type.name.attr)}\\"
}
}
arg101 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = \\"\${type.name.attr}\\"
arg7 = [
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
]
arg8 = \\"\${type.name}\\"
arg9 = [
\\"\${type.name}\\",
\\"\${type.name}\\",
\\"\${type.name}\\",
]
arg10 = [
\\"e\\",
4,
true,
\\"\${type.name.attr}\\",
\\"\${type.name}\\",
]
arg11 = \\"\${arg}\\"
arg12 = \\"\${type.name.attr}\\"
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = \\"\${fn(1, 2, 3)}\\"
arg16 = \\"\${fn(type.name.attr, type.name.attr, type.name.attr)}\\"
arg100 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = \\"\${type.name.attr}\\"
arg7 = [
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
]
arg8 = \\"\${type.name}\\"
arg9 = [
\\"\${type.name}\\",
\\"\${type.name}\\",
\\"\${type.name}\\",
]
arg10 = [
\\"e\\",
4,
true,
\\"\${type.name.attr}\\",
\\"\${type.name}\\",
]
arg11 = \\"\${arg}\\"
arg12 = \\"\${type.name.attr}\\"
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = \\"\${fn(1, 2, 3)}\\"
arg16 = \\"\${fn(type.name.attr, type.name.attr, type.name.attr)}\\"
}
}
arg102 = {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = \\"\${type.name.attr}\\"
arg7 = [
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
]
arg8 = \\"\${type.name}\\"
arg9 = [
\\"\${type.name}\\",
\\"\${type.name}\\",
\\"\${type.name}\\",
]
arg10 = [
\\"e\\",
4,
true,
\\"\${type.name.attr}\\",
\\"\${type.name}\\",
]
arg11 = \\"\${arg}\\"
arg12 = \\"\${type.name.attr}\\"
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = \\"\${fn(1, 2, 3)}\\"
arg16 = \\"\${fn(type.name.attr, type.name.attr, type.name.attr)}\\"
arg100 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = \\"\${type.name.attr}\\"
arg7 = [
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
]
arg8 = \\"\${type.name}\\"
arg9 = [
\\"\${type.name}\\",
\\"\${type.name}\\",
\\"\${type.name}\\",
]
arg10 = [
\\"e\\",
4,
true,
\\"\${type.name.attr}\\",
\\"\${type.name}\\",
]
arg11 = \\"\${arg}\\"
arg12 = \\"\${type.name.attr}\\"
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = \\"\${fn(1, 2, 3)}\\"
arg16 = \\"\${fn(type.name.attr, type.name.attr, type.name.attr)}\\"
}
arg101 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = \\"\${type.name.attr}\\"
arg7 = [
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
]
arg8 = \\"\${type.name}\\"
arg9 = [
\\"\${type.name}\\",
\\"\${type.name}\\",
\\"\${type.name}\\",
]
arg10 = [
\\"e\\",
4,
true,
\\"\${type.name.attr}\\",
\\"\${type.name}\\",
]
arg11 = \\"\${arg}\\"
arg12 = \\"\${type.name.attr}\\"
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = \\"\${fn(1, 2, 3)}\\"
arg16 = \\"\${fn(type.name.attr, type.name.attr, type.name.attr)}\\"
arg100 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = \\"\${type.name.attr}\\"
arg7 = [
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
]
arg8 = \\"\${type.name}\\"
arg9 = [
\\"\${type.name}\\",
\\"\${type.name}\\",
\\"\${type.name}\\",
]
arg10 = [
\\"e\\",
4,
true,
\\"\${type.name.attr}\\",
\\"\${type.name}\\",
]
arg11 = \\"\${arg}\\"
arg12 = \\"\${type.name.attr}\\"
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = \\"\${fn(1, 2, 3)}\\"
arg16 = \\"\${fn(type.name.attr, type.name.attr, type.name.attr)}\\"
}
}
arg101 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = \\"\${type.name.attr}\\"
arg7 = [
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
]
arg8 = \\"\${type.name}\\"
arg9 = [
\\"\${type.name}\\",
\\"\${type.name}\\",
\\"\${type.name}\\",
]
arg10 = [
\\"e\\",
4,
true,
\\"\${type.name.attr}\\",
\\"\${type.name}\\",
]
arg11 = \\"\${arg}\\"
arg12 = \\"\${type.name.attr}\\"
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = \\"\${fn(1, 2, 3)}\\"
arg16 = \\"\${fn(type.name.attr, type.name.attr, type.name.attr)}\\"
arg100 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = \\"\${type.name.attr}\\"
arg7 = [
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
]
arg8 = \\"\${type.name}\\"
arg9 = [
\\"\${type.name}\\",
\\"\${type.name}\\",
\\"\${type.name}\\",
]
arg10 = [
\\"e\\",
4,
true,
\\"\${type.name.attr}\\",
\\"\${type.name}\\",
]
arg11 = \\"\${arg}\\"
arg12 = \\"\${type.name.attr}\\"
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = \\"\${fn(1, 2, 3)}\\"
arg16 = \\"\${fn(type.name.attr, type.name.attr, type.name.attr)}\\"
}
}
arg101 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = \\"\${type.name.attr}\\"
arg7 = [
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
]
arg8 = \\"\${type.name}\\"
arg9 = [
\\"\${type.name}\\",
\\"\${type.name}\\",
\\"\${type.name}\\",
]
arg10 = [
\\"e\\",
4,
true,
\\"\${type.name.attr}\\",
\\"\${type.name}\\",
]
arg11 = \\"\${arg}\\"
arg12 = \\"\${type.name.attr}\\"
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = \\"\${fn(1, 2, 3)}\\"
arg16 = \\"\${fn(type.name.attr, type.name.attr, type.name.attr)}\\"
arg100 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = \\"\${type.name.attr}\\"
arg7 = [
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
\\"\${type.name.attr}\\",
]
arg8 = \\"\${type.name}\\"
arg9 = [
\\"\${type.name}\\",
\\"\${type.name}\\",
\\"\${type.name}\\",
]
arg10 = [
\\"e\\",
4,
true,
\\"\${type.name.attr}\\",
\\"\${type.name}\\",
]
arg11 = \\"\${arg}\\"
arg12 = \\"\${type.name.attr}\\"
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = \\"\${fn(1, 2, 3)}\\"
arg16 = \\"\${fn(type.name.attr, type.name.attr, type.name.attr)}\\"
}
}
}
}

"
`;

exports[`Data 2`] = `
"data \\"type\\" \\"name\\"{
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = type.name.attr
arg7 = [
type.name.attr,
type.name.attr,
type.name.attr,
]
arg8 = type.name
arg9 = [
type.name,
type.name,
type.name,
]
arg10 = [
\\"e\\",
4,
true,
type.name.attr,
type.name,
]
arg11 = arg
arg12 = type.name.attr
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = fn(1, 2, 3)
arg16 = fn(type.name.attr, type.name.attr, type.name.attr)
arg100 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = type.name.attr
arg7 = [
type.name.attr,
type.name.attr,
type.name.attr,
]
arg8 = type.name
arg9 = [
type.name,
type.name,
type.name,
]
arg10 = [
\\"e\\",
4,
true,
type.name.attr,
type.name,
]
arg11 = arg
arg12 = type.name.attr
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = fn(1, 2, 3)
arg16 = fn(type.name.attr, type.name.attr, type.name.attr)
}
arg101 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = type.name.attr
arg7 = [
type.name.attr,
type.name.attr,
type.name.attr,
]
arg8 = type.name
arg9 = [
type.name,
type.name,
type.name,
]
arg10 = [
\\"e\\",
4,
true,
type.name.attr,
type.name,
]
arg11 = arg
arg12 = type.name.attr
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = fn(1, 2, 3)
arg16 = fn(type.name.attr, type.name.attr, type.name.attr)
arg100 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = type.name.attr
arg7 = [
type.name.attr,
type.name.attr,
type.name.attr,
]
arg8 = type.name
arg9 = [
type.name,
type.name,
type.name,
]
arg10 = [
\\"e\\",
4,
true,
type.name.attr,
type.name,
]
arg11 = arg
arg12 = type.name.attr
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = fn(1, 2, 3)
arg16 = fn(type.name.attr, type.name.attr, type.name.attr)
}
}
arg101 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = type.name.attr
arg7 = [
type.name.attr,
type.name.attr,
type.name.attr,
]
arg8 = type.name
arg9 = [
type.name,
type.name,
type.name,
]
arg10 = [
\\"e\\",
4,
true,
type.name.attr,
type.name,
]
arg11 = arg
arg12 = type.name.attr
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = fn(1, 2, 3)
arg16 = fn(type.name.attr, type.name.attr, type.name.attr)
arg100 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = type.name.attr
arg7 = [
type.name.attr,
type.name.attr,
type.name.attr,
]
arg8 = type.name
arg9 = [
type.name,
type.name,
type.name,
]
arg10 = [
\\"e\\",
4,
true,
type.name.attr,
type.name,
]
arg11 = arg
arg12 = type.name.attr
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = fn(1, 2, 3)
arg16 = fn(type.name.attr, type.name.attr, type.name.attr)
}
}
arg101 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = type.name.attr
arg7 = [
type.name.attr,
type.name.attr,
type.name.attr,
]
arg8 = type.name
arg9 = [
type.name,
type.name,
type.name,
]
arg10 = [
\\"e\\",
4,
true,
type.name.attr,
type.name,
]
arg11 = arg
arg12 = type.name.attr
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = fn(1, 2, 3)
arg16 = fn(type.name.attr, type.name.attr, type.name.attr)
arg100 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = type.name.attr
arg7 = [
type.name.attr,
type.name.attr,
type.name.attr,
]
arg8 = type.name
arg9 = [
type.name,
type.name,
type.name,
]
arg10 = [
\\"e\\",
4,
true,
type.name.attr,
type.name,
]
arg11 = arg
arg12 = type.name.attr
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = fn(1, 2, 3)
arg16 = fn(type.name.attr, type.name.attr, type.name.attr)
}
}
arg102 = {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = type.name.attr
arg7 = [
type.name.attr,
type.name.attr,
type.name.attr,
]
arg8 = type.name
arg9 = [
type.name,
type.name,
type.name,
]
arg10 = [
\\"e\\",
4,
true,
type.name.attr,
type.name,
]
arg11 = arg
arg12 = type.name.attr
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = fn(1, 2, 3)
arg16 = fn(type.name.attr, type.name.attr, type.name.attr)
arg100 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = type.name.attr
arg7 = [
type.name.attr,
type.name.attr,
type.name.attr,
]
arg8 = type.name
arg9 = [
type.name,
type.name,
type.name,
]
arg10 = [
\\"e\\",
4,
true,
type.name.attr,
type.name,
]
arg11 = arg
arg12 = type.name.attr
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = fn(1, 2, 3)
arg16 = fn(type.name.attr, type.name.attr, type.name.attr)
}
arg101 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = type.name.attr
arg7 = [
type.name.attr,
type.name.attr,
type.name.attr,
]
arg8 = type.name
arg9 = [
type.name,
type.name,
type.name,
]
arg10 = [
\\"e\\",
4,
true,
type.name.attr,
type.name,
]
arg11 = arg
arg12 = type.name.attr
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = fn(1, 2, 3)
arg16 = fn(type.name.attr, type.name.attr, type.name.attr)
arg100 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = type.name.attr
arg7 = [
type.name.attr,
type.name.attr,
type.name.attr,
]
arg8 = type.name
arg9 = [
type.name,
type.name,
type.name,
]
arg10 = [
\\"e\\",
4,
true,
type.name.attr,
type.name,
]
arg11 = arg
arg12 = type.name.attr
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = fn(1, 2, 3)
arg16 = fn(type.name.attr, type.name.attr, type.name.attr)
}
}
arg101 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = type.name.attr
arg7 = [
type.name.attr,
type.name.attr,
type.name.attr,
]
arg8 = type.name
arg9 = [
type.name,
type.name,
type.name,
]
arg10 = [
\\"e\\",
4,
true,
type.name.attr,
type.name,
]
arg11 = arg
arg12 = type.name.attr
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = fn(1, 2, 3)
arg16 = fn(type.name.attr, type.name.attr, type.name.attr)
arg100 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = type.name.attr
arg7 = [
type.name.attr,
type.name.attr,
type.name.attr,
]
arg8 = type.name
arg9 = [
type.name,
type.name,
type.name,
]
arg10 = [
\\"e\\",
4,
true,
type.name.attr,
type.name,
]
arg11 = arg
arg12 = type.name.attr
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = fn(1, 2, 3)
arg16 = fn(type.name.attr, type.name.attr, type.name.attr)
}
}
arg101 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = type.name.attr
arg7 = [
type.name.attr,
type.name.attr,
type.name.attr,
]
arg8 = type.name
arg9 = [
type.name,
type.name,
type.name,
]
arg10 = [
\\"e\\",
4,
true,
type.name.attr,
type.name,
]
arg11 = arg
arg12 = type.name.attr
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = fn(1, 2, 3)
arg16 = fn(type.name.attr, type.name.attr, type.name.attr)
arg100 {
arg0 = \\"a\\"
arg1 = [
\\"b\\",
\\"c\\",
\\"d\\",
]
arg2 = 0
arg3 = [
1,
2,
3,
]
arg4 = true
arg5 = [
false,
true,
false,
]
arg6 = type.name.attr
arg7 = [
type.name.attr,
type.name.attr,
type.name.attr,
]
arg8 = type.name
arg9 = [
type.name,
type.name,
type.name,
]
arg10 = [
\\"e\\",
4,
true,
type.name.attr,
type.name,
]
arg11 = arg
arg12 = type.name.attr
arg13 = as is
arg14 = <<EOT
heredoc
EOT
arg15 = fn(1, 2, 3)
arg16 = fn(type.name.attr, type.name.attr, type.name.attr)
}
}
}
}

"
`;
